File: internal/ocr/format.go
package ocr

import (
	"bytes"
	"fmt"
	"net/http"
)

type fileKind int

const (
	fileUnknown fileKind = iota
	fileImage
	filePDF
)

func sniffFile(b []byte) fileKind {
	if isPDF(b) {
		return filePDF
	}
	if sniffImage(b) {
		return fileImage
	}
	return fileUnknown
}

func isPDF(b []byte) bool {
	if http.DetectContentType(b) == "application/pdf" {
		return true
	}
	n := len(b)
	if n > 1024 {
		n = 1024
	}
	return bytes.Contains(b[:n], []byte("%PDF-"))
}

// PreviewJPEG turns an uploaded bill into a JPEG for the confirm screen.
// Photos are compressed as before. A PDF stacks every
// rasterized page (up to maxPDFPages) when pdftoppm is available, otherwise
// a text slip.
func PreviewJPEG(raw []byte) ([]byte, error) {
	if len(raw) == 0 {
		return nil, ErrNoImage
	}
	if len(raw) > MaxImageBytes {
		return nil, fmt.Errorf("file must be 10 MB or smaller")
	}
	switch sniffFile(raw) {
	case filePDF:
		return previewPDF(raw)
	case fileImage:
		return PrepareJPEG(raw)
	default:
		return nil, fmt.Errorf("file must be jpeg, png, webp, gif, or pdf")
	}
}

File: src/ocr/format.ts
import { sniffImage } from './image';
import { prepareJPEG } from './image';
import { previewPDF, renderTextSlip } from './pdf';
import { MaxImageBytes, NoImageError } from './types';

export const fileUnknown = 0;
export const fileImage = 1;
export const filePDF = 2;
export type FileKind = typeof fileUnknown | typeof fileImage | typeof filePDF;

export function sniffFile(b: Buffer): FileKind {
  if (isPDF(b)) return filePDF;
  if (sniffImage(b)) return fileImage;
  return fileUnknown;
}

export function isPDF(b: Buffer): boolean {
  if (b.length >= 5 && b.subarray(0, 5).toString('ascii') === '%PDF-') return true;
  const n = Math.min(b.length, 1024);
  return b.subarray(0, n).includes(Buffer.from('%PDF-'));
}

export async function previewJPEG(raw: Buffer): Promise<Buffer> {
  if (raw.length === 0) throw new NoImageError();
  if (raw.length > MaxImageBytes) throw new Error('file must be 10 MB or smaller');
  switch (sniffFile(raw)) {
    case filePDF:
      return previewPDF(raw);
    case fileImage:
      return prepareJPEG(raw);
    default:
      throw new Error('file must be jpeg, png, webp, gif, or pdf');
  }
}

export async function previewText(text: string): Promise<Buffer> {
  return renderTextSlip(text);
}
